adapters/ads_adapter.py

```diff
@@ -1,22 +1,173 @@
 
 """
+from lognplot.client import LognplotTcpClient
+import pyads
+import time
+import ctypes
+import struct
+import datetime
+import re
+import argparse
+from collections import namedtuple
+from typing import Type
 
-# TODO: implement.
 
+class AdsClient:
+    Entry = namedtuple(
+        "Entry", ["name", "typename", "comment", "datatype", "datatype_size"]
+    )
 
+    DATATYPE_MAP = {
+        "BOOL": pyads.PLCTYPE_BOOL,
+        "BYTE": pyads.PLCTYPE_BYTE,
+        "DINT": pyads.PLCTYPE_DINT,
+        "DWORD": pyads.PLCTYPE_DWORD,
+        "INT": pyads.PLCTYPE_INT,
+        "LREAL": pyads.PLCTYPE_LREAL,
+        "REAL": pyads.PLCTYPE_REAL,
+        "SINT": pyads.PLCTYPE_SINT,
+        "UDINT": pyads.PLCTYPE_UDINT,
+        "UINT": pyads.PLCTYPE_UINT,
+        "USINT": pyads.PLCTYPE_USINT,
+        "WORD": pyads.PLCTYPE_WORD,
+    }  # type: Dict[str, Type]
 
-from lognplot.client import LognplotTcpClient
+    def __init__(self, ams_net_id: str, ams_net_port: str):
+        self.notification_handles = []
+        self.plc = pyads.Connection(ams_net_id, ams_net_port)
+        self.plc.open()
+
+    def __del__(self):
+        for handle in self.notification_handles:
+            self.plc.del_device_notification(*handle)
+
+        self.plc.close()
+
+    def log(self, name: str, plc_type):
+        attr = pyads.NotificationAttrib(ctypes.sizeof(plc_type))
+        handles = self.plc.add_device_notification(
+            name, attr, self.callback(plc_type), pyads.ADSTRANS_SERVERCYCLE
+        )
+        self.notification_handles.append(handles)
+
+    def subscribe(self, pattern: str):
+        parsed_entries = self.get_ads_entries()
+
+        for entry in parsed_entries:
+            if re.match(pattern, entry.name) != None:
+                self.log(entry.name, self.DATATYPE_MAP[entry.typename])
+
+    def callback(self, plc_type):
+        @self.plc.notification(plc_type)
+        def decorated_callback(handle, name, timestamp, value):
+            client.send_sample(
+                name, AdsClient.format_timestamp(timestamp), float(value),
+            )
+
+        return decorated_callback
+
+    def get_ads_entries(self):
+        upload_info = self._get_upload_info()
+        entries = self.plc.read(
+            pyads.constants.ADSIGRP_SYM_UPLOAD, 0, ctypes.c_ubyte * upload_info.nSymSize
+        )
+
+        parsed_entries = self._unpack_entries(
+            bytes(entries), upload_info.nSymSize, upload_info.nSymbols
+        )
+        return parsed_entries
+
+    @staticmethod
+    def format_timestamp(timestamp: datetime):
+        return time.mktime(timestamp.timetuple()) + timestamp.microsecond / 1e6
+
+    def _get_upload_info(self):
+        upload_info = self.plc.read(
+            pyads.constants.ADSIGRP_SYM_UPLOADINFO,
+            0,
+            pyads.structs.SAdsSymbolUploadInfo,
+        )
+        return upload_info
+
+    def _unpack_entry(data: bytes, index: int) -> (int, Entry):
```

Comment:
Since this function is unrelated to this class, it might make sense to move it out of the class and make it a simple function.